ord_schema/interface/search.py

```diff
@@ -0,0 +1,126 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Query web interface to the Open Reaction Database in Postgres.
+
+The client is stateless. Full information is in the URL so results can be
+linked.
+
+There are three kinds of query.
+
+  inputs and output SMILES or SMARTS
+
+    These match against individual reagents and results. Each SMILES expression
+    can be freely specified for "exact", "substructure", or "similarity"
+    matching. SMARTS expressions do not allow such a specification.
+
+  reaction ID
+
+    Matches either zero or one reactions. The ID has the form,
+      "ord-<32-bit hex>".
+
+  reaction SMILES
+
+    Matches against RDKit fingerprints.
+
+Query parameters are communicated in URL GET params.
+
+  input=<pattern;(exact|substructure|similarity|smarts)>
+
+  output=<pattern;(exact|substructure|similarity|smarts)>
+
+    The token after the semicolon specifies the matching criterion. The default
+    is "exact". The pattern is a SMILES string, unless the token is "smarts" in
+    which case the pattern is a SMARTS string.
+
+    The "input" param may be repeated. The "output" param may not.
+
+  reaction_id=<id>
+
+  reaction_smiles=<smiles>
+
+If multiple conditions are given, then the query is interpreted as conjunction.
+
+All query parameters are assumed to be URL-encoded.
+"""
+
+import flask
+
+import query
+
+app = flask.Flask(__name__, template_folder='.')
+
+
+@app.route('/')
+def show_root():
+    """Shows the web form.
+
+    If there are query params, then the query is executed and the form is
+    populated with the results. The form fields are populated with the params.
+    """
+    reaction_ids = flask.request.args.get('reaction_ids')
+    reaction_smiles = flask.request.args.get('reaction_smiles')
+    inputs = flask.request.args.getlist("input")
+    output = flask.request.args.get('output')
+    similarity = flask.request.args.get('similarity')
+
+    predicate = query.Predicate()
+
+    if reaction_ids is not None:
+        for reaction_id in reaction_ids.split(','):
+            predicate.add_reaction_id(reaction_id)
+
+    if reaction_smiles is not None:
+        predicate.set_reaction_smiles(reaction_smiles)
+
+    in_splits = [i.split(';', 1) for i in inputs] if inputs else []
+    for smiles, mode_name in in_splits:
+        mode = query.Predicate.MatchMode.from_name(mode_name)
+        predicate.add_input(smiles, mode)
+
+    if output is not None:
+        smiles, mode_name = output.split(';', 1)
+        mode = query.Predicate.MatchMode.from_name(mode_name)
+        predicate.set_output(smiles, mode)
+
+    if similarity is not None:
+        predicate.set_tanimoto_threshold(float(similarity))
+
+    if reaction_ids or reaction_smiles or inputs or output:
+        db = connect()
+        reaction_ids = db.predicate_search_ids(predicate)
```

Comment:
Would it be a good idea to add error handling here? (I typo'd a SMILES string and caused a crash)